File: apps/dispatch-control/lib/galley/spanResolver.ts
/**
 * Per-block quotedSpan resolution (GLY-02).
 *
 * Anchors QA `qaCorrections` findings onto exact character offsets within a
 * single draft block's text -- NEVER the whole section's concatenated text
 * (a cross-block "match" is not a real match; see 32-RESEARCH.md Pattern 2).
 *
 * A wrong highlight is worse than an honest miss (D-12): ambiguity of any
 * kind -- 2+ matches with no usable hint, or still-ambiguous after
 * normalization -- always resolves to `unresolved`, never a guess.
 *
 * Pure, dependency-free.
 */

export interface ResolvedAnnotation {
  findingId: string // Convex _id
  sectionId: string // galley/draft section id (post-mapping)
  blockIndex: number
  start: number // char offset into blocks[blockIndex].text
  end: number
  severity: 'info' | 'warning' | 'error'
  axis?: string
  reason: string
  suggestedFix?: string
  quotedSpan: string
}

export interface UnresolvedFinding {
  findingId: string
  sectionId: string
  severity: 'info' | 'warning' | 'error'
  axis?: string
  reason: string
  suggestedFix?: string
  quotedSpan?: string
}

/** Minimal shape needed from a draft row -- matches `ContentBlock` structurally. */
export interface DraftBlockLike {
  type: string
  text: string
}

/** Minimal shape needed from a Convex `qaCorrections` row, already filtered to one section. */
export interface QaFinding {
  _id: string
  severity: 'info' | 'warning' | 'error'
  axis?: string
  reason: string
  suggestedFix?: string
  quotedSpan?: string
  blockIndexHint?: number
  accepted?: boolean
  /** §33.1 resolution state — callers filter via `isOpenFinding` (findingState.ts). */
  resolution?: 'accepted' | 'dismissed'
}

interface MatchCandidate {
  blockIndex: number
  start: number
  end: number
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

// Curly -> straight quote substitution ONLY -- a 1:1 char swap, so it never
// changes string length. Offsets computed against the normalized string
// therefore map directly onto the ORIGINAL (un-normalized) string.
function normQuotes(s: string): string {
  return s.replace(/[‘’]/g, "'").replace(/[“”]/g, '"')
}

function exactMatchesFor(blocks: DraftBlockLike[], quoted: string): MatchCandidate[] {
  const out: MatchCandidate[] = []
  blocks.forEach((block, blockIndex) => {
    const idx = block.text.indexOf(quoted)
    if (idx >= 0) out.push({ blockIndex, start: idx, end: idx + quoted.length })
  })
  return out
}

// Stage A: quote-substitution normalization (length-preserving). Still
// requires an exact substring match after normalizing both sides.
function quoteNormalizedMatchesFor(blocks: DraftBlockLike[], quoted: string): MatchCandidate[] {
  const normalizedQuoted = normQuotes(quoted)
  const out: MatchCandidate[] = []
  blocks.forEach((block, blockIndex) => {
    const idx = normQuotes(block.text).indexOf(normalizedQuoted)
    if (idx >= 0) out.push({ blockIndex, start: idx, end: idx + quoted.length })
  })
  return out
}

// Stage B: whitespace-tolerant regex over the quote-normalized text. Quote
// substitution never changes length, so `match.index` / `match[0].length`
// index the ORIGINAL block text directly -- never a normalized string.
function whitespaceTolerantMatchesFor(blocks: DraftBlockLike[], quoted: string): MatchCandidate[] {
  const normalizedQuoted = normQuotes(quoted)
  const pattern = escapeRegExp(normalizedQuoted).replace(/\s+/g, '\\s+')
  const out: MatchCandidate[] = []
  blocks.forEach((block, blockIndex) => {
    const re = new RegExp(pattern, 'g')
    const normalizedText = normQuotes(block.text)
    let match: RegExpExecArray | null
    while ((match = re.exec(normalizedText)) !== null) {
      out.push({ blockIndex, start: match.index, end: match.index + match[0].length })
      if (match[0].length === 0) re.lastIndex += 1
    }
  })
  return out
}

/**
 * Resolves a list of match candidates to a single winner, or signals
 * ambiguity. Never guesses: a hint only disambiguates when it names a block
 * that is ACTUALLY one of the candidates; an invalid/stale/non-matching
 * hint is ignored entirely (falls through within the caller's stage
 * sequencing, or is treated as unresolved if this is the final stage).
 */
function disambiguate(
  matches: MatchCandidate[],
  blockIndexHint: number | undefined,
  blockCount: number,
): MatchCandidate | 'ambiguous' | undefined {
  if (matches.length === 0) return undefined
  if (matches.length === 1) return matches[0]
  if (typeof blockIndexHint === 'number' && blockIndexHint >= 0 && blockIndexHint < blockCount) {
    const hinted = matches.find(m => m.blockIndex === blockIndexHint)
    if (hinted) return hinted
  }
  return 'ambiguous'
}

/**
 * Resolves QA findings (already filtered to one section) against that
 * section's draft blocks. Findings are searched one block at a time --
 * never against the blocks joined together.
 */
export function resolveSectionFindings(
  blocks: DraftBlockLike[],
  findings: QaFinding[],
  sectionId: string,
): { resolved: ResolvedAnnotation[]; unresolved: UnresolvedFinding[] } {
  const resolved: ResolvedAnnotation[] = []
  const unresolved: UnresolvedFinding[] = []

  for (const finding of findings) {
    // Defensive: the caller filters out accepted findings, but tolerate one
    // slipping through by excluding it entirely (not even as unresolved).
    if (finding.accepted === true) continue

    const quoted = finding.quotedSpan
    if (!quoted) {
      unresolved.push({
        findingId: finding._id,
        sectionId,
        severity: finding.severity,
        axis: finding.axis,
        reason: finding.reason,
        suggestedFix: finding.suggestedFix,
        quotedSpan: finding.quotedSpan,
      })
      continue
    }

    let candidate = disambiguate(exactMatchesFor(blocks, quoted), finding.blockIndexHint, blocks.length)

    if (candidate === undefined) {
      candidate = disambiguate(
        quoteNormalizedMatchesFor(blocks, quoted),
        finding.blockIndexHint,
        blocks.length,
      )
    }
    if (candidate === undefined) {
      candidate = disambiguate(
        whitespaceTolerantMatchesFor(blocks, quoted),
        finding.blockIndexHint,
        blocks.length,
      )
    }

    if (candidate && candidate !== 'ambiguous') {
      resolved.push({
        findingId: finding._id,
        sectionId,
        blockIndex: candidate.blockIndex,
        start: candidate.start,
        end: candidate.end,
        severity: finding.severity,
        axis: finding.axis,
        reason: finding.reason,
        suggestedFix: finding.suggestedFix,
        quotedSpan: quoted,
      })
    } else {
      unresolved.push({
        findingId: finding._id,
        sectionId,
        severity: finding.severity,
        axis: finding.axis,
        reason: finding.reason,
        suggestedFix: finding.suggestedFix,
        quotedSpan: quoted,
      })
    }
  }

  return { resolved, unresolved }
}

File: packages/pipeline/src/eisenbalm_pipeline/lib/span_resolver.py
"""Server-side quotedSpan resolution (Phase 33, docs/API_CONTRACTS.md §33.5).

A 1:1 Python port of ``apps/dispatch-control/lib/galley/spanResolver.ts`` so
client and server resolution always agree — a finding the galley renders as
anchored must not 409 on accept, and vice versa.

Anchors a QA finding's ``quotedSpan`` onto exact character offsets within a
SINGLE draft block's text — NEVER the whole section's concatenated text (a
cross-block "match" is not a real match). Three stages, tried in order,
each searched block-by-block:

  1. Exact:                ``block["text"].find(quoted)``
  2. Quote-normalized:     curly ``‘’“”`` -> straight ``'"`` on BOTH sides —
                           a 1:1, length-preserving char swap, so offsets
                           computed on normalized text index the ORIGINAL
                           text directly.
  3. Whitespace-tolerant:  ``re.escape`` the quote-normalized span with every
                           whitespace run collapsed to a ``\\s+`` pattern,
                           matched via ``re.finditer`` over the
                           quote-normalized block text; the match's own
                           ``start()``/``end()`` index the original text.

Disambiguation (identical to the TS ``disambiguate``): 0 matches -> next
stage; exactly 1 -> winner; 2+ -> ``block_index_hint`` wins ONLY if it names
an actual candidate block, otherwise ambiguous = unresolved (``None``).

**Never guess** (D-12/D-05): a wrong anchor is worse than an honest miss —
ambiguity of any kind resolves to ``None``, which the accept endpoint maps
to 409 ``span_not_resolved``.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional, Union


@dataclass(frozen=True)
class Match:
    """A resolved span: offsets index blocks[block_index]["text"] directly."""

    block_index: int
    start: int
    end: int


# Sentinel distinguishing "2+ matches, unresolvable" (stop — unresolved)
# from "0 matches" (None — fall through to the next stage).
_AMBIGUOUS = object()

_CURLY_SINGLE = re.compile(r"[‘’]")
_CURLY_DOUBLE = re.compile(r"[“”]")


def _norm_quotes(s: str) -> str:
    """Curly -> straight quote substitution ONLY — a 1:1 char swap, so it
    never changes string length. Offsets computed against the normalized
    string therefore map directly onto the ORIGINAL (un-normalized) string."""
    return _CURLY_DOUBLE.sub('"', _CURLY_SINGLE.sub("'", s))


def _exact_matches(blocks: list[dict], quoted: str) -> list[Match]:
    out: list[Match] = []
    for i, block in enumerate(blocks):
        idx = block.get("text", "").find(quoted)
        if idx >= 0:
            out.append(Match(block_index=i, start=idx, end=idx + len(quoted)))
    return out


def _quote_normalized_matches(blocks: list[dict], quoted: str) -> list[Match]:
    """Stage 2: quote-substitution normalization (length-preserving). Still
    requires an exact substring match after normalizing both sides."""
    normalized_quoted = _norm_quotes(quoted)
    out: list[Match] = []
    for i, block in enumerate(blocks):
        idx = _norm_quotes(block.get("text", "")).find(normalized_quoted)
        if idx >= 0:
            out.append(Match(block_index=i, start=idx, end=idx + len(quoted)))
    return out


def _whitespace_tolerant_matches(blocks: list[dict], quoted: str) -> list[Match]:
    """Stage 3: whitespace-tolerant regex over the quote-normalized text.
    Quote substitution never changes length, so ``m.start()``/``m.end()``
    index the ORIGINAL block text directly — never a normalized string.
    The matched run in the TEXT may differ in length from ``quoted``.

    NOTE: unlike the TS ``escapeRegExp``, Python's ``re.escape`` ALSO escapes
    whitespace chars (``' ' -> '\\ '``), so a naive
    ``re.sub(r"\\s+", ..., re.escape(s))`` would leave orphaned backslashes.
    Split on whitespace runs FIRST, escape each part, then join with
    ``\\s+`` — semantically identical to the TS pattern build."""
    normalized_quoted = _norm_quotes(quoted)
    pattern = r"\s+".join(
        re.escape(part) for part in re.split(r"\s+", normalized_quoted)
    )
    out: list[Match] = []
    for i, block in enumerate(blocks):
        normalized_text = _norm_quotes(block.get("text", ""))
        for m in re.finditer(pattern, normalized_text):
            out.append(Match(block_index=i, start=m.start(), end=m.end()))
    return out


def _disambiguate(
    matches: list[Match],
    block_index_hint: Optional[int],
    block_count: int,
) -> Union[Match, None, object]:
    """Resolve candidates to a single winner, or signal ambiguity.

    Never guesses: a hint only disambiguates when it names a block that is
    ACTUALLY one of the candidates; an invalid/stale/non-matching hint is
    ignored entirely (-> ambiguous -> unresolved at the caller).

    Returns:
        None       — 0 matches (caller falls through to the next stage)
        Match      — exactly one winner
        _AMBIGUOUS — 2+ matches with no usable hint (caller stops: unresolved)
    """
    if not matches:
        return None
    if len(matches) == 1:
        return matches[0]
    if (
        isinstance(block_index_hint, int)
        and not isinstance(block_index_hint, bool)
        and 0 <= block_index_hint < block_count
    ):
        for m in matches:
            if m.block_index == block_index_hint:
                return m
    return _AMBIGUOUS


def resolve_span(
    blocks: list[dict],
    quoted: str,
    block_index_hint: Optional[int] = None,
) -> Optional[Match]:
    """Resolve ``quoted`` against a section's draft blocks ({type,text}[]).

    Searches one block at a time — never against the blocks joined together.
    Returns a :class:`Match` whose offsets index the ORIGINAL block text, or
    ``None`` when the span is missing OR ambiguous (never a guess).
    """
    if not quoted or not blocks:
        return None

    candidate = _disambiguate(
        _exact_matches(blocks, quoted), block_index_hint, len(blocks)
    )
    if candidate is None:
        candidate = _disambiguate(
            _quote_normalized_matches(blocks, quoted), block_index_hint, len(blocks)
        )
    if candidate is None:
        candidate = _disambiguate(
            _whitespace_tolerant_matches(blocks, quoted),
            block_index_hint,
            len(blocks),
        )

    if candidate is None or candidate is _AMBIGUOUS:
        return None
    return candidate
